src/generate-types.ts

```diff
@@ -261,1117 +83,1396 @@ const statusCodes = [
+      return ts.createImportDeclaration(
         undefined,
-        ts.createNamespaceImport(ts.createIdentifier(external.importAs))
-      ),
-      ts.createStringLiteral(external.importFile)
-    );
-  });
-}
-
-function generateBuiltins(runtimeModule: string) {
-  return ts.createNodeArray([
-    ts.createImportDeclaration(
-      undefined,
-      undefined,
-      ts.createImportClause(undefined, ts.createNamespaceImport(runtimeLibrary)),
-      ts.createStringLiteral(runtimeModule)
-    )
-  ]);
-}
+        undefined,
+        ts.createImportClause(
+          undefined,
+          ts.createNamespaceImport(ts.createIdentifier(external.importAs))
+        ),
+        ts.createStringLiteral(external.importFile)
+      );
+    });
+  }
 
-interface ImportDefinition {
-  importAs: string;
-  importFile: string;
-}
-export interface Options {
-  externalOpenApiImports: readonly ImportDefinition[];
-  externalOpenApiSpecs: (url: string) => string | undefined;
-  oas: oas.OpenAPIObject;
-  runtimeModule: string;
-  emitStatusCode: (status: number) => boolean;
-  unsupportedFeatures?: {
-    security?: UnsupportedFeatureBehaviour;
-  };
-}
+  function generateBuiltins() {
+    return ts.createNodeArray([
+      ts.createImportDeclaration(
+        undefined,
+        undefined,
+        ts.createImportClause(undefined, ts.createNamespaceImport(runtimeLibrary)),
+        ts.createStringLiteral(options.runtimeModule)
+      )
+    ]);
+  }
 
-function addIndexSignatureIgnores(src: string) {
-  const result: string[] = [];
-  src.split('\n').forEach(line => {
-    const m = line.match(new RegExp('\\[\\s*' + valueClassIndexSignatureKey));
-    if (m) {
-      if (!/\b(unknown|any)\b/.test(line)) {
-        result.push('// @ts-ignore tsc does not like the branding type in index signatures');
+  function addIndexSignatureIgnores(src: string) {
+    const result: string[] = [];
+    src.split('\n').forEach(line => {
+      const m = line.match(new RegExp('\\[\\s*' + valueClassIndexSignatureKey));
+      if (m) {
+        if (!/\b(unknown|any)\b/.test(line)) {
+          result.push('// @ts-ignore tsc does not like the branding type in index signatures');
+          result.push(line);
+          return;
+        }
+      }
+      const brandMatch = line.match(new RegExp('\\s*private readonly ' + oatsBrandFieldName));
+      if (brandMatch) {
+        result.push('// @ts-ignore tsc does not like unused privates');
         result.push(line);
         return;
       }
-    }
-    const brandMatch = line.match(new RegExp('\\s*private readonly ' + oatsBrandFieldName));
-    if (brandMatch) {
-      result.push('// @ts-ignore tsc does not like unused privates');
       result.push(line);
-      return;
+    });
+    return result.join('\n');
+  }
+
+  function addToImports(
+    importAs: string,
+    importFile: string | undefined,
+    action?: (() => Promise<void>) | undefined
+  ) {
+    if (!state.imports[importAs]) {
```

Comment:
this maintains the import list, actions to run state and the global state of which files we are already going to generate as we are bit loose on the promise waiting part.